agent/lib/kontena/network_adapters/weave.rb

```diff
@@ -76,25 +81,81 @@ def already_started?
       @started == true
     end
 
+    def cleanup_ipam
+      loop do
+        sleep 60 * 3 # 3mins
+        debug "starting ipam cleanup process"
+        ipam_container = Docker::Container.get('kontena-ipam-plugin') rescue nil
+        if ipam_container
+          local_addresses = []
+          Docker::Container.all(all: true).each do |container|
+            local_addresses << container.overlay_cidr
+          end
+          cmd = ['/app/bin/kontena-ipam-cleanup'] + local_addresses.compact
+          debug "executing cleanup with command: #{cmd}"
+          ipam_container.exec(cmd) { |stream, chunk| debug "#{stream}: #{chunk}" }
+        end
+      end
+    end
+
     # @param [Hash] opts
     def modify_create_opts(opts)
+      ensure_weave_wait
+
+      image = Docker::Image.get(opts['Image'])
+      image_config = image.info['Config']
+      cmd = []
+      if opts['Entrypoint']
+        if opts['Entrypoint'].is_a?(Array)
+          cmd = cmd + opts['Entrypoint']
+        else
+          cmd = cmd + [opts['Entrypoint']]
+        end
+      end
+      if !opts['Entrypoint'] && image_config['Entrypoint'] && image_config['Entrypoint'].size > 0
+        cmd = cmd + image_config['Entrypoint']
+      end
+      if opts['Cmd'] && opts['Cmd'].size > 0
+        if opts['Cmd'].is_a?(Array)
+          cmd = cmd + opts['Cmd']
+        else
+          cmd = cmd + [opts['Cmd']]
+        end
+      elsif image_config['Cmd'] && image_config['Cmd'].size > 0
+        cmd = cmd + image_config['Cmd']
+      end
+      opts['Entrypoint'] = ['/w/w']
+      opts['Cmd'] = cmd
+
       modify_host_config(opts)
+      opts
+    end
+
+    # @param [Hash] opts
+    def modify_network_opts(opts)
+      opts['Labels']['io.kontena.container.overlay_cidr'] = @ipam_client.reserve_address('kontena')
```

Comment:
The reference to `@ipam_client` needs to be synchronized with the `on_ipam_start`:

```
kontena-agent | E, [2016-11-10T15:09:29.173299 #1] ERROR -- : Actor crashed!
kontena-agent | NoMethodError: undefined method `reserve_address' for nil:NilClass
kontena-agent | 	/app/lib/kontena/network_adapters/weave.rb:136:in `modify_network_opts'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/call/sync.rb:16:in `dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
kontena-agent | 	/usr/lib/ruby/gems/2.3.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```